src/plugins/intel_cpu/src/nodes/common/cpu_convert.cpp

```diff
@@ -893,6 +893,48 @@ struct ConvertFrom4BitPrecision<std::tuple<src_t, dst_t>> {
     }
 };
 
+#define INTEL_CPU_CVT_TO_4BIT_LIST                                              \
+    INTEL_CPU_CVT(f32, nf4), INTEL_CPU_CVT(f16, nf4), INTEL_CPU_CVT(bf16, nf4)
+
+struct ConvertTo4BitContext {
+    ov::element::Type_t outType;
+    const void* srcPtr;
+    void* dstPtr;
+    size_t size;
+    bool converted;
+};
+
+template <typename T>
+struct ConvertTo4BitPrecision;
+
+template <typename src_t, typename dst_t>
+struct ConvertTo4BitPrecision<std::tuple<src_t, dst_t>> {
+    void operator()(ConvertTo4BitContext& ctx) {
+        auto insert_half_byte = [](uint8_t dst, uint8_t val, bool high_half) -> uint8_t {
+            uint8_t shift = high_half ? 4 : 0;
+            return dst | (uint8_t) (val << shift);
+        };
+
+        auto src = static_cast<const src_t*>(ctx.srcPtr);
+        auto dst = static_cast<uint8_t*>(ctx.dstPtr);
+        // each byte must be fully processed within same thread
+        auto work_amount = div_up(ctx.size, 2);
```

Comment:
On 64 bit platforms if the work_amount is odd, there shouldn't be a problem, as the address is at least 64bit aligned so the last element will always be located in a an allocated page as the maxim data size is 4 bytes. But on 32bit systems accessing [ib*2+1] index may (theoretically of course) hit unallocated page. If we expect that this code may be run on 32 bit systems, probably it makes sense to process the last odd index outside the loop.

Reply:
Good catch. Totally agree. Refactored